rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/locations/impl/ArrayLineOffsetMap.java

```diff
@@ -0,0 +1,153 @@
+    private final IntArray lines;
+    private final ArrayList<IntArray> wideColumnOffsets;
+
+    public ArrayLineOffsetMap(IntArray lines, ArrayList<IntArray> wideColumnOffsets) {
+        this.lines = lines;
+        this.wideColumnOffsets = wideColumnOffsets;
+    }
+
+    @Override
+    public int translateColumn(int line, int column, boolean isEnd) {
+        int lineIndex = lines.search(line);
+        if (lineIndex < 0) {
+            return column;
+        }
+        IntArray lineOffsets = wideColumnOffsets.get(lineIndex);
+        int columnIndex = lineOffsets.search(column);
+        if (columnIndex >= 0) {
+            // exact hit
+            if (isEnd) {
+                // for a end cursor, we want to count this char twice as well
+                columnIndex++;
+            }
+        }
+        else {
+            columnIndex = Math.abs(columnIndex + 1);
+        }
+        // now we know how many
+        return column + columnIndex;
+    }
+
+
+    @SuppressWarnings("java:S3776") // parsing tends to be complex
+    public static LineColumnOffsetMap build(String contents) {
+        int line = 0;
+        int column = 0;
+        char prev = '\0';
+        GrowingIntArray linesWithSurrogate = new GrowingIntArray();
+        ArrayList<IntArray> linesMap = new ArrayList<>(0);
+        GrowingIntArray currentLine = new GrowingIntArray();
+
+        for(int i = 0, n = contents.length() ; i < n ; i++) {
+            char c = contents.charAt(i);
+            if (c == '\n' || c == '\r') {
+                if (c != prev && (prev == '\r' || prev == '\n')) {
+                    continue; // multichar newline skip it
+                }
+                if (!currentLine.isEmpty()) {
+                    linesWithSurrogate.add(line);
+                    linesMap.add(currentLine.build());
+                    currentLine = new GrowingIntArray();
+                }
+                line++;
+                column = 0;
+            }
+            else {
+                column++;
+                if (Character.isHighSurrogate(c) && (i + 1) < n && Character.isLowSurrogate(contents.charAt(i + 1))) {
+                    // full surrogate pair, register it, and skip the next char
+                    currentLine.add(column);
+                    i++;
+                }
+            }
+            prev = c;
+        }
+        if (!currentLine.isEmpty()) {
+            // handle last line
+            linesWithSurrogate.add(line);
+            linesMap.add(currentLine.build());
+        }
+        if (linesMap.isEmpty()) {
+            return EMPTY_MAP;
+        }
+        return new ArrayLineOffsetMap(linesWithSurrogate.build(), linesMap);
+    }
+
+    private static LineColumnOffsetMap EMPTY_MAP = new LineColumnOffsetMap(){
+        @Override
+        public int translateColumn(int line, int column, boolean atEnd) {
+            return column;
+        }
+    };
+
+
+    private static class GrowingIntArray {
```

Comment:
was ArrayList not good enough?

Reply:
We are going to be creating a lot of these, and I want to avoid boxing, and allocating more than needed.

The code is small, it's just a dynamically growing array, instead of putting this inline with the code, I just factored it out to a class.